sphinx/registry.py

```diff
@@ -98,10 +109,19 @@ def has_domain(self, domain):
     def create_domains(self, env):
         # type: (BuildEnvironment) -> Iterator[Domain]
         for DomainClass in itervalues(self.domains):
-            yield DomainClass(env)
+            domain = DomainClass(env)
+
+            # transplant components added by extensions
+            domain.object_types.update(self.domain_object_types.get(domain.name, {}))
```

Comment:
OMG, `domain.objtypes_for_role` ignores this updates...
I will try this again later.